src/avatar.cpp

```diff
@@ -1142,3 +1143,118 @@ void avatar::vomit()
     }
     Character::vomit();
 }
+
+int avatar::get_str_base() const
+{
+    return Character::get_str_base() + std::max( 0, str_upgrade );
+}
+
+int avatar::get_dex_base() const
+{
+    return Character::get_dex_base() + std::max( 0, dex_upgrade );
+}
+
+int avatar::get_int_base() const
+{
+    return Character::get_int_base() + std::max( 0, int_upgrade );
+}
+
+int avatar::get_per_base() const
+{
+    return Character::get_per_base() + std::max( 0, per_upgrade );
+}
+
+int avatar::kill_xp() const
+{
+    // TODO: game::kills probably should be avatar::kills
+    return g->kill_xp();
+}
+
+// based on  D&D 5e level progression
+static const std::vector<int> xp_cutoffs = {
+    300, 900, 2700, 6500, 14000,
+    23000, 34000, 48000, 64000, 85000,
+    100000, 120000, 140000, 165000, 195000,
+    225000, 265000, 305000, 355000, 405000
+};
+
+int avatar::free_upgrade_points() const
+{
+    const int xp = kill_xp();
+    int lvl = 0;
+    for( const int &xp_lvl : xp_cutoffs ) {
+        if( xp >= xp_lvl ) {
+            lvl++;
+        } else {
+            break;
+        }
+    }
+    return lvl - str_upgrade - dex_upgrade - int_upgrade - per_upgrade;
+}
+
+static int xp_to_next()
```

Comment:
To go along with a change to `array<T, N>` from `vector<T>` we can simplify the logic in here a little.
```cpp
const int cur_xp = g->u.kill_xp();
// Initialize to 'already max level' sentinel
int xp_next = -1;
// Iterate in reverse: { 405k, 355k, 305k, ..., 300 }
for( std::array<int, 20>::const_reverse_iterator iter = xp_cutoffs.crbegin();
    iter != xp_cutoffs.crend(); ++iter )
{
    if( cur_xp < *iter ) {
        xp_next = *iter;
    }
}
return xp_next;
```
Simplifies the logic while maintaining the output requirement of finding the lowest experience threshold greater than current experience